OfflineService/OfflineServiceViewController.swift

```diff
@@ -0,0 +1,311 @@
+                    print("Unpacked \(unpackedBytes) of \(totalBytes) bytes")
+                } completionHandler: { (numberOfTiles, error) in
+                    do {
+                        if FileManager.default.fileExists(atPath: temporaryPackURL.path) {
+                            try FileManager.default.removeItem(at: temporaryPackURL)
+                        }
+                    } catch {
+                        self.tilesUnpackingLock.unlock()
+                        
+                        print("Failed to remove temporary pack archive. Error: \(error)")
+                        self.presentAlert(OfflineServiceConstants.title, message: error.localizedDescription)
+                    }
+                    
+                    self.tilesUnpackingLock.unlock()
+                    print("Finished unpacking \(numberOfTiles) tiles")
+                }
+            } catch {
+                tilesUnpackingLock.unlock()
+                
+                print("Error occured while unpacking navigation tiles: \(error)")
+                presentAlert(OfflineServiceConstants.title, message: error.localizedDescription)
+            }
+        }
+    }
+    
+    public func onExpired(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata, pack: OfflineDataPack) {
+        print("[OfflineServiceObserver] \(#function)")
+    }
+    
+    public func onErrored(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata, pack: OfflineDataPack) {
+        print("[OfflineServiceObserver] \(#function)")
+    }
+    
+    public func onDeleting(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata, pack: OfflineDataPack, callback: @escaping OfflineDataPackAcknowledgeCallback) {
+        print("[OfflineServiceObserver] \(#function)")
+        
+        presentAlert(OfflineServiceConstants.title, message: "Would you like to remove \(metadata.id)?", handler: { _ in
+            callback()
+        })
+    }
+    
+    public func onDeleted(for domain: OfflineDataDomain, metadata: OfflineDataRegionMetadata) {
+        print("[OfflineServiceObserver] \(#function)")
+        
+        updateOfflineDataRegions(for: domain, metadata: metadata, delete: true)
+    }
+    
+    public func onInitialized() {
+        print("[OfflineServiceObserver] \(#function)")
+    }
+    
+    public func onIdle() {
+        print("[OfflineServiceObserver] \(#function)")
+    }
+    
+    public func onLogMessage(forMessage message: String) {
+        print("[OfflineServiceObserver] \(#function): \(message)")
+        
+        presentAlert(title, message: message)
+    }
+    
+    public var peer: MBXPeerWrapper?
+    
+    // MARK: - UIViewController delegate methods
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        setupUI()
+        listAvailableRegions()
+    }
+    
+    deinit {
+        // FIXME: In some cases this call might cause crash.
```

Comment:
looks like crash happens when you try to unregister object which wasn't registered

Reply:
That's correct, on the other hand there is clash in data which is retrieved from `OfflineService.listAvailableRegions(forCallback:)` and `OfflineServiceObserver.onAvailable(for:metadata:pack:)` methods and previously I was calling first one and only in case if data was received observer was created. In case of error observer wasn't created - this lead to the crash. In any case I'd expect `OfflineServiceObserver` not to crash in such case, probably something we'd like to flag to Core SDK team.